keysign/discover.py

```diff
@@ -28,13 +28,27 @@ def __init__(self, userdata, discovery):
             self.discovery = AvahiKeysignDiscoveryWithMac()
         self.bt = None
         self.stopped = False
+    
+    def success_callback(self, response):
+        #If the we find key with the given userdata, load the key
+        log.debug("key_data sucessfuly loaded")
+
+    def failure_callback(self, error):
+        #if we get an execption due to failure log the message
+        log.debug("malformed or altered key key received, loading keydata failed ")
 
     @inlineCallbacks
     def start(self):
         # First we try Avahi, if it fails we fallback to Bluetooth because
         # the receiver may be able to use only one of them
         log.info("Trying to use this code with Avahi: %s", self.userdata)
-        key_data = yield threads.deferToThread(self.discovery.find_key, self.userdata)
+        
+        #chain callback to handle failure gracefully and prevent from crash
+        data = threads.deferToThread(self.discovery.find_key, self.userdata)
+        data.addCallback(self.success_callback)
+        data.addErrback(self.failure_callback)
+        key_data = yield data
```

Comment:
It would be better to avoid mixing the current inlineCallbacks with addCallback and addErrback.
The Twisted and Pythonic way to handle an exception with inlineCallbacks is by just surrounding the `deferToThread` with a `try ... except` block.

Reply:
Ahhh, I thought about the try....except block, I'll get this fixed too.

Reply:
@RyuzakiKK   I included it in try....catch block 

try:
    key_data = yield threads.deferToThread(self.discovery.find_key, self.userdata)
except Exception as e:
    log.debug('Unable to get key_data')
    returnValue(e)

but it logs the following output on receiving a malformed key.

```
keysign.util (DEBUG): finished downloading 100 bytes
keysign.gpgmks (DEBUG): Building cmd: 'gpg' '--command-fd' '0' '--with-fingerprint' '--list-options' 'show-sig-subpackets,show-uid-validity,show-unusable-uids,show-unusable-subkeys,show-keyring,show-sig-expire' '--status-fd' '2' '--quiet' '--batch' '--fixed-list-mode' '--no-tty' '--with-colons' '--use-agent' '--homedir' '/tmp/pygpg-ShkJWJ' '--import'
keysign.discover (DEBUG): Unable to get key_data
Unhandled error in Deferred:
```

Reply:
I'm unable right now to test it myself, so I'm not 100% sure that what I say is true.
I suppose that you get an `Unhandled error` because of the `returnValue(e)`. 
The caller is still waiting for a tuple with 3 arguments, this one here https://github.com/gnome-keysign/gnome-keysign/blob/master/keysign/discover.py#L52

Reply:
@RyuzakiKK  Thank you, that was exactly the reason for which it was throwing an unhandled error.
It's working now.
```
try:
    key_data = yield threads.deferToThread(self.discovery.find_key, self.userdata)
    log.debug("Received key successfully")
except Exception as e:
    log.debug('Unable to get key_data, received malformed or altered key')
    key_data = None
    success = False
    message = ""
    returnValue((key_data, success, message))
```
Is this looks good?

Reply:
LGTM

Reply:
Great, I've updated the PR.